### PR TITLE
feat(validation): enforce replay regression policy gates (#229)

### LINKED ISSUE #229: V-06: Baseline replay and validation gates for merge/release
Parent: #223
Owner lane: Team V4 + Team F (Reliability)
Depends on: #225, #226, #227, #228

## Goal
Use trader-validated baselines to enforce regression policy in CI and release flow.

## Scope
1. Baseline promotion flow from completed validation runs.
2. Replay flow comparing current output vs approved baseline.
3. Policy engine for blocking decisions:
   - merge gate,
   - release gate.
4. Hard-fail rules and configurable metric drift tolerances.
5. CI integration and required check wiring.

## Required tests
1. Replay contract tests with pass/fail scenarios.
2. CI gating tests proving block behavior on regressions.
3. Threshold configuration tests.

## Definition of done
1. Baseline replay runs in CI.
2. Regression failures block merge and release.
3. Policy decisions are persisted and auditable.

## Non-goals
1. Human review UI.

## Status updates
Post plan + every PR update in #223.

### PR DESCRIPTION
**Pull request by iamtxena**
## Summary

Implement issue #229 by enforcing baseline replay comparison and policy-driven merge/release blocking gates for validation regressions.

## Related Issue

Fixes #229
Parent: #223
Depends on: #228 (already merged into `main`)

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Infrastructure/CI changes

## Changes Made

- Added replay policy evaluator module to compute:
  - replay decision matrix (`pass` / `conditional_pass` / `fail`)
  - threshold breach detection from baseline vs candidate drift deltas
  - merge/release blocked booleans and CI gate statuses
- Updated `/v2/validation-regressions/replay` runtime behavior to:
  - enforce completed baseline/candidate runs
  - enforce same-strategy replay comparisons
  - return detailed replay policy outputs (gate status, threshold fields, reasons)
- Extended v2 contract schema/OpenAPI and regenerated SDK model for replay response fields.
- Added contract/integration tests for required coverage:
  - replay pass/fail matrix
  - threshold behavior
  - CI gate behavior
  - endpoint-level gate payload assertions

## Testing

Describe how you tested these changes:

- [ ] Ran `bun lint` (frontend)
- [ ] Ran `bun typecheck` (frontend)
- [x] Ran `uv run pytest` (backend)
- [x] Tested manually in development environment
- [x] Added new tests for this feature/fix

Executed checks:

- `uv run --with pytest --with pytest-asyncio pytest tests/contracts/test_validation_replay_policy.py tests/contracts/test_platform_api_v2_handlers.py -k 'replay'`
- `uv run --with pytest pytest tests/contracts/test_openapi_contract_v2_baseline.py`
- `uv run --with pytest pytest tests/contracts/test_openapi_contract_v2_validation_freeze.py tests/contracts/test_runtime_openapi_contract_v2.py tests/contracts/test_sdk_validation_contract_shape.py`
- `uv run --with pytest --with pytest-asyncio pytest tests/contracts/test_validation_schema_contract.py tests/contracts/test_sdk_validation_contract_shape.py tests/contracts/test_openapi_contract_v2_validation_freeze.py tests/contracts/test_openapi_contract_v2_baseline.py tests/contracts/test_runtime_openapi_contract_v2.py tests/contracts/test_platform_api_v2_handlers.py -k 'validation or replay'`
- `bash contracts/scripts/mock-smoke-test.sh`
- `bash contracts/scripts/mock-consumer-contract-test.sh`
- `python3 contracts/scripts/check-slo-alert-baseline.py`
- `bash contracts/scripts/verify-sdk-drift.sh`

## Screenshots (if applicable)

N/A

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my code
- [x] I have commented my code where necessary
- [x] I have updated documentation as needed
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged
- [ ] I confirmed all PR review threads are resolved before requesting merge
- [ ] For Platform API/runtime contract changes, Cursor and Greptile reviews are completed

---

> [!NOTE]
> **High Risk**
> Changes a public v2 contract (new required replay fields) and adds CI gates that can block merges/releases based on replay evaluation and new persistence/migrations, so misconfiguration or policy logic bugs could stop deployments or break clients.
> 
> **Overview**
> Adds **policy-driven replay regression gates** to `/v2/validation-regressions/replay`: replays now require *completed* baseline/candidate runs from the same `strategyId`, are evaluated immediately via new `evaluate_replay_policy`, and return/persist detailed results (merge/release block flags + gate status, baseline/candidate decisions, drift delta vs threshold, and machine-readable `reasons`).
> 
> Introduces persistence for replay outcomes (`ValidationReplayMetadata`, new `validation_replays` Supabase table + tenant-aware RLS migration) and wires it into both in-memory and Supabase storage adapters.
> 
> Updates the OpenAPI spec + regenerated TypeScript SDK model to include the new required replay fields, and adds CI enforcement by running `python -m src.platform_api.validation.release_gate_check` in contracts governance and backend deploy workflows, with new contract/integration tests covering schema, storage, and gate behavior.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 2d41551604622e60e0e22320e41d9f54dd4fe8c7. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

<h3>Greptile Summary</h3>

Implements baseline replay comparison with policy-driven merge/release gates for validation regressions. The PR adds a new `validation_replay_policy.py` module that evaluates replay outcomes based on decision regression (comparing baseline vs candidate using a ranking system) and metric drift threshold checks. The `/v2/validation-regressions/replay` endpoint now enforces stricter preconditions (both runs must be `completed` and share the same `strategyId`) and returns detailed gate information including blocking status, threshold breach detection, and machine-readable reasons.

## Key Changes

- New replay policy evaluator computes decision matrix (`pass`/`conditional_pass`/`fail`) based on candidate-to-baseline regression and drift delta exceeding threshold
- Updated endpoint returns comprehensive gate payload: `mergeBlocked`, `releaseBlocked`, gate statuses, baseline/candidate decisions, drift metrics, and reasons
- Extended v2 contract schema with new required fields in `ValidationRegressionReplay` response
- Regenerated TypeScript SDK model with proper type safety for new fields
- Comprehensive test coverage including decision matrix, threshold boundary behavior, and integration tests

## Potential Concerns

- Breaking change: Existing clients expecting the old minimal replay response shape will break when consuming the new response with additional required fields
- The metric drift calculation uses `max(0.0, candidate - baseline)`, which means baseline improvements (lower candidate drift) don't trigger threshold checks - verify this is intentional policy behavior

<h3>Confidence Score: 4/5</h3>

- This PR is generally safe to merge with careful consideration of the breaking changes to the public API contract
- Score reflects high-quality implementation with excellent test coverage and clean separation of concerns, but deducted one point due to the breaking contract changes that will impact existing SDK consumers and the need to verify drift calculation semantics
- backend/src/platform_api/services/validation_v2_service.py:729-741 requires verification that the drift delta calculation using max(0.0, candidate - baseline) aligns with product requirements

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| backend/src/platform_api/services/validation_replay_policy.py | New module implementing replay policy evaluation logic with decision ranking, threshold checking, and gate blocking - implementation is clean and well-tested |
| backend/src/platform_api/services/validation_v2_service.py | Integrated replay policy evaluator into `/v2/validation-regressions/replay` endpoint with proper precondition checks and comprehensive response fields |
| backend/src/platform_api/schemas_v2.py | Extended `ValidationRegressionReplay` schema with new gate status, decision, and threshold fields - properly validated with Pydantic constraints |
| backend/tests/contracts/test_validation_replay_policy.py | New comprehensive test suite covering decision matrix, threshold boundary behavior, and CI gate status - excellent coverage of edge cases |
| backend/tests/contracts/test_platform_api_v2_handlers.py | Added integration tests for replay endpoint asserting new response fields and policy-driven gate behavior - validates end-to-end functionality |

</details>

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    Start[Client requests replay comparison] --> ValidateInput[Validate baseline and candidate IDs]
    ValidateInput --> CheckStatus{Both runs completed?}
    CheckStatus -->|No| Error1[Error: Runs must be completed]
    CheckStatus -->|Yes| CheckStrategy{Same strategyId?}
    CheckStrategy -->|No| Error2[Error: Different strategies]
    CheckStrategy -->|Yes| GatherInputs[Gather policy inputs:<br/>baseline/candidate decisions,<br/>metric drift percentages,<br/>threshold, block flags]
    
    GatherInputs --> EvalPolicy[evaluate_replay_policy]
    
    EvalPolicy --> CheckRegression{Candidate decision<br/>rank less than baseline?}
    CheckRegression -->|Yes| SetFail1[decision = fail<br/>Add reason]
    CheckRegression -->|No| KeepDecision[decision = candidate_decision]
    
    SetFail1 --> CalcDrift
    KeepDecision --> CalcDrift[Calculate drift delta]
    
    CalcDrift --> CheckThreshold{drift_delta greater than threshold?}
    CheckThreshold -->|Yes| SetFail2[decision = fail<br/>Add threshold reason]
    CheckThreshold -->|No| CheckDecision
    SetFail2 --> CheckDecision
    
    CheckDecision{Final decision?}
    CheckDecision -->|fail| BlockFail[Apply block_merge_on_fail<br/>and block_release_on_fail]
    CheckDecision -->|conditional_pass| BlockAgent[Apply block_merge_on_agent_fail<br/>and block_release_on_agent_fail]
    CheckDecision -->|pass| NoBlock[No blocking]
    
    BlockFail --> BuildResponse
    BlockAgent --> BuildResponse
    NoBlock --> BuildResponse[Build ValidationRegressionReplay response]
    
    BuildResponse --> Return[Return 202 with replay payload]
```

<sub>Last reviewed commit: b59cc43</sub>

### CODE REVIEW

File: backend/src/platform_api/services/validation_replay_policy.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    candidate_rank = _DECISION_RANK[inputs.candidate_decision]
+    baseline_rank = _DECISION_RANK[inputs.baseline_decision]
+    decision: ValidationReplayDecision = inputs.candidate_decision
+    reasons: list[str] = []
+
+    if candidate_rank < baseline_rank:
+        decision = "fail"
+        reasons.append("candidate_decision_regressed_from_baseline")
+
+    metric_drift_delta_pct = max(0.0, candidate_metric_drift_pct - baseline_metric_drift_pct)
```
[the remark this answers is not in the document]
Comment (iamtxena) [reply backend/src/platform_api/services/validation_replay_policy.py]: Confirmed and intentional for regression policy: threshold applies to *worsening* drift relative to baseline, not improvements. Added an explicit inline comment and dedicated contract test in commit 91db2bf (`test_replay_policy_threshold_ignores_candidate_metric_improvement`).

File: backend/src/platform_api/services/validation_v2_service.py [hunk 2]
```diff
@@ -687,29 +691,78 @@ async def replay_validation_regression(
...
+        else:
+            summary = "Replay comparison passed without regression."
         replay = ValidationRegressionReplay(
             id=replay_id,
             baselineId=request.baselineId,
             candidateRunId=request.candidateRunId,
             status="queued",
             decision=decision,
-            summary="Replay accepted for execution.",
+            mergeBlocked=replay_outcome.merge_blocked,
```
[the remark this answers is not in the document]
Comment (iamtxena) [reply backend/src/platform_api/services/validation_v2_service.py]: Addressed in commit 91db2bf: replay status now returns `completed` because the policy evaluation is computed synchronously before response construction (`backend/src/platform_api/services/validation_v2_service.py`). Added handler assertions to lock this behavior (`backend/tests/contracts/test_platform_api_v2_handlers.py`).

**Comment by iamtxena** [comment]
Addressed the remaining CONDITIONAL_PASS gap from #229 reviewer notes.

New commit: 2d41551

What changed:
- Added merge-time replay gate enforcement to required PR check workflow:
  - .github/workflows/contracts-governance.yml
  - step: `python -m src.platform_api.validation.release_gate_check`
- Added contract guardrail test to prevent regression:
  - backend/tests/contracts/test_validation_ci_gate_contract.py
  - asserts replay preflight exists in contract-governance (PR path) and backend-deploy.

Targeted verification:
- test_validation_ci_gate_contract.py
- test_validation_release_gate_check.py
- test_validation_replay_policy.py
(all passing locally)

This closes the specific reviewer concern that replay gating was only post-merge (deploy flow).
